File: app/src/main/java/com/boardgamegeek/ui/ImageActivity.java
package com.boardgamegeek.ui;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.BitmapDrawable;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.graphics.Palette;
import android.text.TextUtils;
import android.view.View;
import android.widget.ImageView;

import com.boardgamegeek.R;
import com.boardgamegeek.util.HttpUtils;
import com.boardgamegeek.util.PaletteTransformation;
import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.answers.ContentViewEvent;
import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import butterknife.BindView;
import butterknife.ButterKnife;
import timber.log.Timber;

public class ImageActivity extends AppCompatActivity {
	private static final String KEY_IMAGE_URL = "IMAGE_URL";

	@BindView(R.id.image) ImageView imageView;
	@BindView(R.id.progress) View progressBar;

	public static void start(Context context, String imageUrl) {
		if (TextUtils.isEmpty(imageUrl)) {
			Timber.w("Missing the required image URL.");
			return;
		}
		Intent intent = new Intent(context, ImageActivity.class);
		intent.putExtra(KEY_IMAGE_URL, imageUrl);
		context.startActivity(intent);
	}

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		setContentView(R.layout.activity_image);
		ButterKnife.bind(this);

		final String imageUrl = getIntent().getStringExtra(KEY_IMAGE_URL);
		if (TextUtils.isEmpty(imageUrl)) {
			Timber.w("Received an empty imageUrl");
			finish();
			return;
		}

		if (savedInstanceState == null) {
			String imageId = Uri.parse(imageUrl).getLastPathSegment();
			Answers.getInstance().logContentView(new ContentViewEvent()
				.putContentType("Image")
				.putContentId(imageId));
		}

		Picasso.with(this)
			.load(HttpUtils.ensureScheme(imageUrl))
			.error(R.drawable.thumbnail_image_empty)
			.fit()
			.centerInside()
			.transform(PaletteTransformation.instance())
			.into(imageView, new Callback.EmptyCallback() {
				@Override
				public void onSuccess() {
					Bitmap bitmap = ((BitmapDrawable) imageView.getDrawable()).getBitmap();
					Palette palette = PaletteTransformation.getPalette(bitmap);
					imageView.setBackgroundColor(getColor(palette));
					progressBar.setVisibility(View.GONE);
				}
			});
	}

	private int getColor(Palette palette) {
		Palette.Swatch swatch = palette.getDarkMutedSwatch();
		if (swatch == null) {
			swatch = palette.getDarkVibrantSwatch();
		}
		if (swatch == null) {
			swatch = palette.getMutedSwatch();
		}
		if (swatch == null) {
			return Color.BLACK;
		}
		return swatch.getRgb();
	}
}

File: app/src/main/java/com/boardgamegeek/ui/ImageActivity.kt
package com.boardgamegeek.ui

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.net.toUri
import androidx.core.view.isVisible
import com.boardgamegeek.R
import com.boardgamegeek.databinding.ActivityImageBinding
import com.boardgamegeek.extensions.ensureHttpsScheme
import com.boardgamegeek.extensions.startActivity
import com.boardgamegeek.util.PaletteTransformation
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.logEvent
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso
import timber.log.Timber

class ImageActivity : AppCompatActivity() {
    private lateinit var firebaseAnalytics: FirebaseAnalytics
    private lateinit var binding : ActivityImageBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        firebaseAnalytics = FirebaseAnalytics.getInstance(this)

        binding = ActivityImageBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val imageUrl = intent.getStringExtra(KEY_IMAGE_URL)
        if (imageUrl.isNullOrBlank()) {
            Timber.w("Received an empty imageUrl")
            finish()
            return
        }

        if (savedInstanceState == null) {
            firebaseAnalytics.logEvent(FirebaseAnalytics.Event.VIEW_ITEM) {
                param(FirebaseAnalytics.Param.CONTENT_TYPE, "Image")
                param(FirebaseAnalytics.Param.ITEM_ID, imageUrl.toUri().lastPathSegment.orEmpty())
            }
        }

        Picasso.with(this)
            .load(imageUrl.ensureHttpsScheme())
            .error(R.drawable.thumbnail_image_empty)
            .fit()
            .centerInside()
            .transform(PaletteTransformation.instance())
            .into(binding.imageView, object : Callback.EmptyCallback() {
                override fun onSuccess() {
                    setBackgroundColor()
                    binding.progressBar.isVisible = false
                }
            })
    }

    private fun setBackgroundColor() {
        val bitmap = (binding.imageView.drawable as BitmapDrawable).bitmap
        val palette = PaletteTransformation.getPalette(bitmap)
        val swatch = palette?.darkMutedSwatch ?: palette?.darkVibrantSwatch ?: palette?.mutedSwatch
        binding.imageView.setBackgroundColor(swatch?.rgb ?: Color.BLACK)
    }

    companion object {
        private const val KEY_IMAGE_URL = "IMAGE_URL"

        fun start(context: Context, imageUrl: String?) {
            if (imageUrl.isNullOrBlank()) {
                Timber.w("Missing the required image URL.")
                return
            }
            context.startActivity<ImageActivity>(KEY_IMAGE_URL to imageUrl)
        }
    }
}
